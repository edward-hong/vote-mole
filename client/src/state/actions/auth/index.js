import { FETCH_USER, FETCH_USER_FULFILLED } from '../../../constants'

export const fetchUser = () => ({ type: FETCH_USER })

export const fetchUserFulfilled = payload => ({
	type: FETCH_USER_FULFILLED,
	payload,
})
