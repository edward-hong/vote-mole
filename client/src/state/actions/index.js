export * from './auth'
export * from './polls'
