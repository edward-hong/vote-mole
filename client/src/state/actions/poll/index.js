import {
	POLL_INFO_TYPE,
	POLL_INFO_RECEIVED_TYPE,
	POLL_CLEAR_TYPE,
} from '../../../constants'

// Fetch poll
export const fetchPoll = payload => ({
	type: POLL_INFO_TYPE,
	payload,
})

// Dispatch received poll data to reducer
export const dispatchPollReceived = payload => ({
	type: POLL_INFO_RECEIVED_TYPE,
	payload,
})

// Action for clearing poll data from state
export const clearPoll = () => ({ type: POLL_CLEAR_TYPE })
