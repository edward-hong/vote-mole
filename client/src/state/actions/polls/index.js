import {
	POLL_ALL_TYPE,
	POLL_ALL_RECEIVED_TYPE,
	POLL_CLEAR_TYPE,
	POLL_USER_TYPE,
	POLL_USER_RECEIVED_TYPE,
} from '../../../constants'

// Fetch all polls
export const fetchAllPolls = payload => ({
	type: POLL_ALL_TYPE,
	payload,
})

// Dispatch received all polls to reducer
export const dispatchAllPollsReceived = payload => ({
	type: POLL_ALL_RECEIVED_TYPE,
	payload,
})

// Fetch user polls
export const fetchUserPolls = payload => ({
	type: POLL_USER_TYPE,
	payload,
})

// Dispatch received user polls to reducer
export const dispatchUserPollsReceived = payload => ({
	type: POLL_USER_RECEIVED_TYPE,
	payload,
})

// Clear polls state
export const clearPolls = () => ({ type: POLL_CLEAR_TYPE })
