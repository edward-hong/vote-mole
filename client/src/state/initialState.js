const initialState = {
	auth: null,
	poll: null,
}

export default initialState
