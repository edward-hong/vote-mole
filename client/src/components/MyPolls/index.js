import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import { Container, Col } from 'reactstrap'
import { dec, multiply, is } from 'ramda'

import Polls from '../Polls'
import PageIndex from '../PageIndex'
import Row from '../styled/Row'
import MainHeading from '../styled/MainHeading'
import { fetchUserPolls, clearPolls } from '../../state/actions'

const MyPolls = ({ auth, polls, fetchUserPolls, clearPolls }) => {
	const pageSize = 5
	const [page, setPage] = useState(1)

	useEffect(() => {
		// Only fetch for user polls when logged in
		if (is(Object, auth)) {
			fetchUserPolls({ id: auth._id, limit: pageSize, offset: 0 })
		}

		return () => {
			clearPolls()
		}
	}, [fetchUserPolls, clearPolls, auth])

	// On page change, change page state and fetch for more data
	const onPageChange = page => {
		setPage(page)
		fetchUserPolls({
			id: auth._id,
			limit: pageSize,
			offset: multiply(pageSize, dec(page)),
		})
	}

	// Only render when logged in
	return is(Object, auth) ? (
		<Container fluid>
			<MainHeading>My Polls</MainHeading>
			<Row>
				<Col
					lg={{ size: '6', offset: 3 }}
					md={{ size: '8', offset: 2 }}
					sm={{ size: '10', offset: 1 }}
					xs="12">
					<Polls {...polls} />
					<PageIndex
						{...polls}
						page={page}
						pageSize={pageSize}
						onPageChange={onPageChange}
					/>
				</Col>
			</Row>
		</Container>
	) : null
}

const mapStateToProps = ({ polls }) => ({ polls })

export default connect(
	mapStateToProps,
	{ fetchUserPolls, clearPolls }
)(MyPolls)
