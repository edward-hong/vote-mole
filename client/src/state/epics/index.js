import { combineEpics } from 'redux-observable'
import { values } from 'ramda'

import * as authEpics from './auth'
import * as pollEpics from './polls'

const rootEpic = combineEpics(...values(authEpics), ...values(pollEpics))

export default rootEpic
