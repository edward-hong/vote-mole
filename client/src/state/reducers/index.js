import { combineReducers } from 'redux'
import { reducer as form } from 'redux-form'

import auth from './auth'
import polls from './polls'

const rootReducer = combineReducers({ auth, form, polls })

export default rootReducer
