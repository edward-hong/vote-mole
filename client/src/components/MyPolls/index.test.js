import React from 'react'
import { fireEvent } from '@testing-library/react'

import MyPolls from '.'
import renderWithRouterAndRedux from '../../utils/test/renderWithRouterAndRedux'
import { fetchUserPolls, clearPolls } from '../../state/actions'

jest.mock('../../state/actions', () => ({
	fetchUserPolls: jest.fn(),
	clearPolls: jest.fn(),
}))

fetchUserPolls.mockImplementation(() => ({ type: 'test' }))
clearPolls.mockImplementation(() => ({ type: 'test' }))

const polls = [
	{
		id: '5cdaf28cee882077685935e3',
		pollQuestion: 'PHP, Python or Ruby?',
	},
	{
		id: '5cdaf0d4ee882077685935de',
		pollQuestion: 'Favourite JavaScript Framework',
	},
	{
		id: '5cdaf0a6ee882077685935d9',
		pollQuestion: 'Favourite Sport',
	},
	{
		id: '5cdaf088ee882077685935d6',
		pollQuestion: 'Favourite Movie',
	},
	{
		id: '5cdaf04fee882077685935d1',
		pollQuestion: 'Favourite Superhero',
	},
	{
		id: '5cdaf04fee882077685djx0qe',
		pollQuestion: 'Favourite Colour',
	},
]

const auth = {
	_id: '5cd19682498a98c6962f34c1',
	profileID: '111881712224957728911',
}

describe('MyPolls', () => {
	it('returns null if `auth` props is false', () => {
		const { container } = renderWithRouterAndRedux(<MyPolls auth={false} />)

		expect(container.firstChild).toBeNull()
	})

	it('renders correctly', () => {
		const { container } = renderWithRouterAndRedux(<MyPolls auth={auth} />, {
			initialState: { polls: { count: polls.length, polls } },
		})

		expect(container.firstChild).toMatchSnapshot()
	})

	it('clicking on the pagination buttons loads new pages', () => {
		const { getByText, container } = renderWithRouterAndRedux(<MyPolls />, {
			initialState: { polls: { count: polls.length, polls }, auth },
		})
		const nextPageButton = getByText('â€º')

		fireEvent.click(nextPageButton)

		expect(container.firstChild).toMatchSnapshot()
	})
})
