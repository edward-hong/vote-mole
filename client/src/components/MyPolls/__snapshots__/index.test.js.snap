// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPolls clicking on the pagination buttons loads new pages 1`] = `
.c3 {
  height: 3.5rem;
  padding: 0;
  border-color: #21252a;
}

.c4 {
  display: block;
  margin: 0;
  width: 100%;
  height: 100%;
  text-align: center;
  color: #21252a;
  line-height: 3.5rem;
}

.c4:hover {
  color: white;
  background-color: #21252a;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.15s ease-in-out,color 0.15s ease-in-out;
  transition: background-color 0.15s ease-in-out,color 0.15s ease-in-out;
}

.c2 {
  height: 276px;
}

.c5 .pagination {
  margin-top: 2rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5 .page-link {
  color: #21252a;
  border-color: #21252a;
}

.c5 .page-link:focus,
.c5 .page-link:hover {
  color: white;
  border-color: #21252a;
  background-color: rgba(33,37,41,0.7);
  -webkit-transition: background-color 0.15s ease-in-out,color 0.15s ease-in-out;
  transition: background-color 0.15s ease-in-out,color 0.15s ease-in-out;
}

.c5 .page-item.active .page-link {
  background-color: #21252a;
  border-color: #21252a;
}

.c1 {
  margin-top: 1.5rem;
}

.c0 {
  text-align: center;
}

<div
  class="container-fluid"
>
  <h1
    class="c0"
  >
    My Polls
  </h1>
  <div
    class="c1 row"
  >
    <div
      class="col-12 col-sm-10 offset-sm-1 col-md-8 offset-md-2 col-lg-6 offset-lg-3"
    >
      <ul
        class="c2 list-group"
      >
        <li
          class="c3 list-group-item"
        >
          <a
            class="c4"
            href="/poll/5cdaf28cee882077685935e3"
          >
            PHP, Python or Ruby?
          </a>
        </li>
        <li
          class="c3 list-group-item"
        >
          <a
            class="c4"
            href="/poll/5cdaf0d4ee882077685935de"
          >
            Favourite JavaScript Framework
          </a>
        </li>
        <li
          class="c3 list-group-item"
        >
          <a
            class="c4"
            href="/poll/5cdaf0a6ee882077685935d9"
          >
            Favourite Sport
          </a>
        </li>
        <li
          class="c3 list-group-item"
        >
          <a
            class="c4"
            href="/poll/5cdaf088ee882077685935d6"
          >
            Favourite Movie
          </a>
        </li>
        <li
          class="c3 list-group-item"
        >
          <a
            class="c4"
            href="/poll/5cdaf04fee882077685935d1"
          >
            Favourite Superhero
          </a>
        </li>
        <li
          class="c3 list-group-item"
        >
          <a
            class="c4"
            href="/poll/5cdaf04fee882077685djx0qe"
          >
            Favourite Colour
          </a>
        </li>
      </ul>
      <div
        class="c5"
      >
        <ul
          class="pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              «
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              ‹
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              1
            </a>
          </li>
          <li
            class="page-item active"
          >
            <a
              class="page-link"
              href="#"
            >
              2
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              ›
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              href="#"
            >
              »
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`MyPolls renders correctly 1`] = `null`;
